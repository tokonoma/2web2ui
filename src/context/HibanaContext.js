import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { isAccountUiOptionSet } from 'src/helpers/conditions/account';
import { selectCondition } from 'src/selectors/accessConditionState';

const HibanaStateContext = createContext();

function Provider(props) {
  const { isHibanaEnabled: _isHibanaEnabled, children, ...rest } = props;

  // TODO: When OG theme is removed, the env variables will no longer be necessary
  const isHibanaEnabled =
    process.env.REACT_APP_DEFAULT_TO_HIBANA === 'true' ? true : _isHibanaEnabled;

  return (
    <HibanaStateContext.Provider value={{ isHibanaEnabled, ...rest }}>
      {children}
    </HibanaStateContext.Provider>
  );
}

function mapStateToProps(state) {
  return {
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
    hasThemeControls: selectCondition(isAccountUiOptionSet('hibana.hasThemeControls'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export function useHibana() {
  const context = useContext(HibanaStateContext);

  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function HibanaConsumer({ children }) {
  return <HibanaStateContext.Consumer>{children}</HibanaStateContext.Consumer>;
}
