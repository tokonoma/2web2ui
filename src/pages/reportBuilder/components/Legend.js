import React from 'react';
import { connect } from 'react-redux';
import { Inline, Tag, Box } from 'src/components/matchbox';

import { selectFeatureFlaggedMetrics } from 'src/selectors/metrics';
import getConfig from 'src/helpers/getConfig';

function renderMetric(metric, uniqueLabel, removeMetric) {
  return (
    <Tag key={metric.name} onRemove={() => removeMetric(metric.name)}>
      <Box as="span" display="inline-flex" alignItems="center">
        <Box //The color circle
          display="inline-block"
          height="16px"
          width="16px"
          borderRadius="circle"
          backgroundColor={metric.stroke}
          position="relative"
          right="9px"
        />
        <Box>
          {metric.isUniquePerTimePeriod && uniqueLabel
            ? `${metric.label} ${uniqueLabel}`
            : metric.label}
        </Box>
      </Box>
    </Tag>
  );
}

export const Legend = props => {
  const { metrics, featureFlaggedMetrics, reportOptions, removeMetric } = props;
  const { useMetricsRollup } = featureFlaggedMetrics;
  const { precision } = reportOptions;

  const metricsRollupPrecisionMap = getConfig('metricsRollupPrecisionMap');
  const precisionObj = metricsRollupPrecisionMap.find(p => p.value === precision);
  return (
    <Box marginTop="100">
      {metrics && (
        <Inline space="300">
          <Box fontWeight="600">Metrics </Box>
          {metrics.map(metric =>
            renderMetric(
              metric,
              precisionObj && useMetricsRollup ? precisionObj.uniqueLabel : '',
              removeMetric,
            ),
          )}
        </Inline>
      )}
    </Box>
  );
};

const mapStateToProps = state => ({
  reportOptions: state.reportOptions,
  featureFlaggedMetrics: selectFeatureFlaggedMetrics(state),
});

export default connect(mapStateToProps)(Legend);
