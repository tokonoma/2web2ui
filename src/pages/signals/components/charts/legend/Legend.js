import React from 'react';
import { Tooltip } from 'src/components/matchbox';
import { useHibana } from 'src/context/HibanaContext';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import OGStyles from './Legend.module.scss';
import hibanaStyles from './LegendHibana.module.scss';

const Item = ({ label, OGFill, hibanaFill, tooltipContent }) => {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;
  const styles = useHibanaOverride(OGStyles, hibanaStyles);
  const fill = isHibanaEnabled ? hibanaFill : OGFill;

  const content = (
    <div className={styles.Item}>
      <span className={styles.Fill} style={{ background: fill }} />
      <span className={styles.Label}>{label}</span>
    </div>
  );

  if (tooltipContent) {
    return (
      <Tooltip content={tooltipContent(label)} dark>
        {content}
      </Tooltip>
    );
  }

  return content;
};

const Legend = ({ items, tooltipContent }) => (
  <div>
    {items.map((item, i) => (
      <Item key={i} tooltipContent={tooltipContent} {...item} />
    ))}
  </div>
);

export default Legend;
