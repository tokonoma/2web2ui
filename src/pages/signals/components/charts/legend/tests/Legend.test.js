import React from 'react';
import { shallow } from 'enzyme';
import { useHibana } from 'src/context/HibanaContext';
import Legend from '../Legend';

jest.mock('src/context/HibanaContext');
jest.mock('src/hooks/useHibanaOverride', () => styles => styles);

describe('Legend Component', () => {
  const defaultProps = {
    items: [{ fill: 'blue', label: 'label 1' }, { label: 'label 2' }],
    OGFill: 'whitesmoke',
    hibanaFill: 'blue',
  };

  const subject = props => {
    useHibana.mockImplementationOnce(() => [{ isHibanaEnabled: false }]);

    return shallow(<Legend {...defaultProps} {...props} />);
  };

  it('renders correctly', () => {
    const wrapper = subject();

    expect(wrapper).toMatchSnapshot();
    wrapper.find('Item').forEach(item => {
      expect(item).toMatchSnapshot();
    });
  });

  it('renders correctly with tooltip content', () => {
    const wrapper = subject({ tooltipContent: label => label });
    expect(wrapper).toMatchSnapshot();
    wrapper.find('Item').forEach(item => {
      expect(item).toMatchSnapshot();
    });
  });
});
