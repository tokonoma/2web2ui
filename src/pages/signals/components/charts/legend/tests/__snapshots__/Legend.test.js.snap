// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend Component renders correctly 1`] = `
<div>
  <Item
    fill="blue"
    key="0"
    label="label 1"
  />
  <Item
    key="1"
    label="label 2"
  />
</div>
`;

exports[`Legend Component renders correctly 2`] = `
<Item
  fill="blue"
  key="0"
  label="label 1"
/>
`;

exports[`Legend Component renders correctly 3`] = `
<Item
  key="1"
  label="label 2"
/>
`;

exports[`Legend Component renders correctly with tooltip content 1`] = `
<div>
  <Item
    fill="blue"
    key="0"
    label="label 1"
    tooltipContent={[Function]}
  />
  <Item
    key="1"
    label="label 2"
    tooltipContent={[Function]}
  />
</div>
`;

exports[`Legend Component renders correctly with tooltip content 2`] = `
<Item
  fill="blue"
  key="0"
  label="label 1"
  tooltipContent={[Function]}
/>
`;

exports[`Legend Component renders correctly with tooltip content 3`] = `
<Item
  key="1"
  label="label 2"
  tooltipContent={[Function]}
/>
`;
